flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/PartitionComputer.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.filesystem;
+
+import org.apache.flink.annotation.Internal;
+
+import java.io.Serializable;
+
+/**
+ * Compute partition path from record and project non-partition columns for output writer.
+ *
+ * <p>See {@link RowPartitionComputer}.
+ *
+ * @param <T> The type of the consumed records.
+ */
+@Internal
+public interface PartitionComputer<T> extends Serializable {
+
+	/**
+	 * Compute partition path from record.
+	 *
+	 * @param in input record.
+	 * @return partition path.
+	 */
+	String computePartition(T in) throws Exception;
```

Comment:
returning LinkedHashMap would be better? This class should focus on how to extract partition columns or project out partition columns.

Reply:
There may be some escape character logical in `PartitionComputer`. I am not sure we can unify hive and flink. (Maybe it is better to invoke hive util to generate partition path, so I let string to be return value)

Reply:
You just brought up another reason why you should change the return type. If Flink and Hive have different strategy to construct partition file path, it should be accomplished by:
1. first extract all partition key values from data 
2. then apply some partition path constructor interface

Current method will mix in two kinds of functionality into one interface and hard to extend. For example, hive and Flink will have same logic about how to project out partition columns from input data, aka your another method `projectColumnsToWrite`, right?

Reply:
Provide a big method, implementation can do what want to do.

But yes, we can split this method to:
- `LinkedHashMap makePartitionValues(T)`
- `String makePartitionPath(LinkedHashMap)`
What we need do is: `computer.makePartitionPath(computer.makePartitionValues(row))`.

Reply:
Could you separate the interfaces? I can only see one reasonable reason for `PartitionComputer` to be an interface is we want to make it template to support `Row` and `BaseRow`. But generate a partition path looks like should be other's responsibility, not this interface.

Reply:
Yes, we could, my little concern is that the interfaces are too more... Actually it is not so complex. `PartitionComputer` for Row or BaseRow, they both are simple. We can rely on util in implementations.
With multi interface, that will limit the arguments of implementation. For `LinkedHashMap`, maybe with BaseRow, we can have a higher performance implementation which not rely on `LinkedHashMap` and JAVA Strings.
But here, I agree with you to separate it.